components/style/matching.rs

```diff
@@ -907,23 +900,38 @@ pub trait MatchMethods : TNode {
             context.local_context().applicable_declarations_cache.borrow_mut();
 
         let (damage, restyle_result) = {
-            let mut data_ref = self.mutate_data().unwrap();
-            let mut data = &mut *data_ref;
-            let final_style =
-                self.cascade_node_pseudo_element(context, parent_style,
+            // Compute the parameters for the cascade.
+            let mut old_style = self.get_existing_style();
+            let cacheable = match old_style {
+                None => true,
+                Some(ref mut old) => {
```

Comment:
Hmm... I see the `cacheable` thing now. It's kind of unfortunate we're moving this out of `cascade_node_pseudo_element`, which is conceptually "everything you need to do to get an element style. Is there an easy way to move it back in? Haven't thought too hard about it, probably it's not a big deal if not.